src/librustc_typeck/check/mod.rs

```diff
@@ -4918,18 +4918,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             format!("{} type parameter{}", n, if n == 1 { "" } else { "s" })
         };
 
-        // Check provided type parameters.
-        let type_defs = segment.map_or(&[][..], |(_, generics)| {
-            if generics.parent.is_none() {
-                &generics.types[generics.has_self as usize..]
-            } else {
-                &generics.types
-            }
-        });
-        let required_len = type_defs.iter().take_while(|d| !d.has_default).count();
-        if types.len() > type_defs.len() {
-            let span = types[type_defs.len()].span;
-            let expected_text = count_type_params(type_defs.len());
+        // Check provided parameters.
+        let (ty_non_def_req_len, ty_req_len, lt_req_len) =
+            segment.map_or((0, 0, 0), |(_, generics)| {
+                let param_counts = generics.param_counts();
+
+                let type_params_offset
```

Comment:
I'd call this `has_self` (or better, `own_self` - if you can find other such `let has_self` places you could change those too) just to keep things simple.